src/pack/make_pack_descriptor.cpp

```diff
@@ -0,0 +1,110 @@
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#include <memory>
+#include <regex>
+#include <set>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include "mesh/meshblock.hpp"
+#include "pack/make_pack_descriptor.hpp"
+
+namespace parthenon {
+namespace impl {
+
+PackDescriptor MakePackDescriptorBase(StateDescriptor *psd,
+                                      const std::vector<std::string> &vars,
+                                      const std::vector<bool> &use_regex,
+                                      const std::vector<MetadataFlag> &flags,
+                                      const std::set<PDOpt> &options) {
+  const std::string cache_label{"normal"};
+  using PDCache = PackDescCache<std::vector<std::string>, std::vector<bool>,
+                                std::vector<MetadataFlag>, std::set<PDOpt>>;
+  auto optional_pd =
+      PDCache::CheckForKey(psd, cache_label, vars, use_regex, flags, options);
+  if (optional_pd) return *optional_pd;
+
+  PARTHENON_REQUIRE(vars.size() == use_regex.size(),
+                    "Vargroup names and use_regex need to be the same size.");
+  auto selector = [&](int vidx, const VarID &id, const Metadata &md) {
+    if (flags.size() > 0) {
+      for (const auto &flag : flags) {
+        if (!md.IsSet(flag)) return false;
+      }
+    }
+
+    if (use_regex[vidx]) {
+      if (std::regex_match(std::string(id.label()), std::regex(vars[vidx]))) return true;
+    } else {
+      if (vars[vidx] == id.label()) return true;
+      if (vars[vidx] == id.base_name && id.sparse_id != InvalidSparseID) return true;
+    }
+    return false;
+  };
+
+  auto pd = PackDescriptor(psd, vars, selector, options);
+  PDCache::CachePackDescriptor(psd, cache_label, pd, vars, use_regex, flags, options);
+  return pd;
+}
+
+PackDescriptor MakePackDescriptorBase(StateDescriptor *psd,
+                                      const std::vector<Uid_t> &var_ids,
+                                      const std::vector<MetadataFlag> &flags,
+                                      const std::set<PDOpt> &options) {
+  const std::string cache_label{"uid"};
+  using PDCache =
+      PackDescCache<std::vector<Uid_t>, std::vector<MetadataFlag>, std::set<PDOpt>>;
+  auto optional_pd = PDCache::CheckForKey(psd, cache_label, var_ids, flags, options);
+  if (optional_pd) return *optional_pd;
+
+  auto selector = [&](int vidx, const VarID &id, const Metadata &md) {
+    if (flags.size() > 0) {
```

Comment:
This check is not needed (in combination with the following iterator `for` loop), is it?